### PR TITLE
SPARKC-242: Added ability to build against unreleased Spark versions

### PR DESCRIPTION
**Pull request by jacek-lewandowski**
Set the Spark version to <branch_name>-SNAPSHOT or to v<x.y.z> and the build
 script will automatically fetch Spark from GIT at that version, build it and install
 you the local Maven repository. Finally it will be used. Works also with IDEA.

### CODE REVIEW

File: project/SparkInstaller.scala [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  private lazy val isWindows = sys.props("os.name").toLowerCase().contains("windows")
+
+  private def getAndInstallSpark(version: String, scalaVersion: String, force: Boolean, rootDir: Path): Unit = {
+    if ((force || !isInstalled(version, scalaVersion)) && !attempts.contains(version -> scalaVersion)) {
+      attempts.putIfAbsent(version -> scalaVersion, {})
+      if (isWindows) {
+        println(s"""
+          Cannot automatically fetch and install Spark on Windows. You need to do this by your own.
```
Comment (RussellSpitzer) [inline project/SparkInstaller.scala]: by -> on , you may also want to add "hahahaaha windows"

File: project/SparkInstaller.scala [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    val GitBranchPattern = """(master|branch).*\-SNAPSHOT""".r
+
+    providedVersion match {
+      case GitTagPattern() => true
+      case GitBranchPattern(_) => true
+      case _ => false
+    }
+  }
+
+  private object StdProcessLogger extends sbt.ProcessLogger {
```
Comment (RussellSpitzer) [inline project/SparkInstaller.scala]: Why are we wrapping the logger here?

**Comment by RussellSpitzer** [comment]
Lgtm

File: project/SparkInstaller.scala (hunk 1 above)
Comment (jacek-lewandowski) [inline project/SparkInstaller.scala]: :wink:

File: project/SparkInstaller.scala (hunk 2 above)
Comment (jacek-lewandowski) [inline project/SparkInstaller.scala]: This is used for super-nice Scala process invocations with `!`. I wanted to redirect output streams to the current stdout and stderr, especially for long running tasks, like `mvn install`.

**Comment by RussellSpitzer** [comment]
+1

**Pull request merged**